src/compiler/program.ts

```diff
@@ -3074,6 +3074,28 @@ namespace ts {
                 return;
             }
 
+            const existingResolutions = file.resolvedTypeReferenceDirectiveNames;

```

Comment:
I think you want to add method that does something similar to `` instead of calling `resolveTypeReferenceDirectiveNamesWorker` where the resolutions are reused and call it from `tryReuseStructureFromOldProgram` as well

This change currently has two points lacking:
1. it doesnt handle case where source files are modified and  program tries to compute new module resolutions and type reference directives. (look in `tryReuseStructureFromOldProgram`)
2. Doesnt report error `esolution_modes_are_only_supported_when_moduleResolution_is_node12_or_nodenext` when reusing

Reply:
> it doesnt handle case where source files are modified and program tries to compute new module resolutions and type reference directives.

`tryReuseStructureFromOldProgram` already overwrites `resolvedTypeReferenceDirectiveNames` with new resolutions when they change (and sets `StructureIsReused.SafeModules` rather than `StructureIsReused.Completely` - not that it matters here, since we have a full new resolved resolution list already made in the former case, so don't need to redo what `tryReuseStructureFromOldProgram`'s already calculated) - we just didn't use the result here.

> Doesnt report error esolution_modes_are_only_supported_when_moduleResolution_is_node12_or_nodenext when reusing

Done.

Reply:
`StructureIsReused.SafeModules` can be set even without computing new type reference directive resolutions (https://github.com/microsoft/TypeScript/blob/main/src/compiler/program.ts#L1799)   Thats why you need something like https://github.com/microsoft/TypeScript/blob/main/src/compiler/program.ts#L1483 and https://github.com/microsoft/TypeScript/blob/main/src/compiler/program.ts#L1525